<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>












  


  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  <title>IDEA User's Guide</title>
  <meta name="author" content="Amy Egan">













  
  
  
  
  
  
  
  
  
  
  
  
  <meta http-equiv="Content-Type" content="text/html; CHARSET=iso-iso-8859-1">













  
  
  
  
  
  
  
  
  
  
  
  
  <style type="text/css"><!--
BODY { font-family: serif }
H1 { font-family: sans-serif }
H2 { font-family: sans-serif }
H3 { font-family: sans-serif }
H4 { font-family: sans-serif }
H5 { font-family: sans-serif }
H6 { font-family: sans-serif }
SUB { font-size: smaller }
SUP { font-size: smaller }
PRE { font-family: monospace }
--></style>
  
  
  
  
  
  
  
  
  
  
  
  
  <meta http-equiv="CONTENT-TYPE" content="text/html; charset=utf-8">













  
  
  
  
  
  
  
  
  
  
  
  
  <meta name="GENERATOR" content="OpenOffice.org 2.0 (Linux)">













  
  
  
  
  
  
  
  
  
  
  
  
  <meta name="AUTHOR" content="Amy Egan">













  
  
  
  
  
  
  
  
  
  
  
  
  <meta name="CREATED" content="20070212;16045100">













  
  
  
  
  
  
  
  
  
  
  
  
  <meta name="CHANGED" content="16010101;0">













  
  
  
  
  
  
  
  
  
  
  
  
  <style>
<!--
@page { size: 8.5in 11in; margin: 0.79in }
P { margin-bottom: 0.08in }
-->
  </style>
</head>


<body style="background-color: rgb(255, 255, 255);">













<h1 style="text-align: center;"></h1>













<br>













<h1 style="text-align: center;">IDEA</h1>













<h3 style="color: rgb(96, 0, 160); text-align: center;">Interactive
Display for
Evolutionary
Analyses</h3>













<h4 style="text-align: center;">Version 2.5 (December 21, 2009)</h4>













<h4 style="text-align: center;">Documentation Updated November 29, 2010</h4>













<hr>
<h1><a name="CONTENTS">Table
of Contents</a></h1>













<br>













<b><a href="#1">1 Introduction</a></b>
<br>













<br>













<b><a href="#2">2
Running IDEA</a></b>
<br>













<br>













<b><a href="#3">3 Performing an
Analysis</a></b>
<ul>













  <li><a href="#3_1">3.1 Expected Input</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  <ul>













    <li><a href="#3_1_1">3.1.1 Alignment
File(s)</a></li>













    <li><a href="#3_1_2">3.1.2 Tree File(s)</a></li>













    <li><a href="#3_1_3">3.3 Dataset List</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  </ul>













  <li><a href="#3_2">3.2&nbsp;Setting
Parameters</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  <ul>













    <li><a href="#3_2_1">3.2.1 Saving
Configuration</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  </ul>













  
  
  
  
  
  
  
  
  
  
  
  
  <ul>













    <li><a href="#3_2_2">3.2.2 Loading
Configuration</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  </ul>













  <li><a href="#3_3">3.3 Monitoring
Progress</a></li>













  <li><a href="#3_4">3.4 Troubleshooting</a></li>













</ul>













<b><a href="#4">4
Viewing Results</a></b>
<ul>













  <li><a href="#4_1">4.1 Data Table</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  <ul>













    <li><a href="#4_1_1">4.1.1 Histograms</a></li>













    <li><a href="#4_1_2">4.1.2 Detailed
Output</a></li>













    <li><a href="#4_1_3">4.1.3 Likelihood Ratio Tests</a></li>













    <li><a href="#4_1_4">4.1.4 Saving the Table</a></li>













  
  
  
  
  
  
  
  
  
  
  
  
  </ul>













  <li><a href="#4_2">4.2 Selected Sites</a></li>













  <li><a href="#4_3">4.3 Phylogenetic Trees</a></li>













  <li><a href="#4_4">4.4 Results Not Obtained Using
IDEA</a></li>











  <li><a href="#4_5">4.5 Examining Flat Files</a></li>













</ul>













<b><a href="#5">5 New in IDEA 2.4</a></b>
<br>













<br>













<b><a href="#6">6 Further Informatio</a></b><b><a href="external_guide.html#4">n</a></b>
<ul>













  <li><a href="external_guide.html#6_1">6.1 How to
Cite IDEA</a></li>













</ul>













<br>













<hr>
<h1><a name="1">1 Introduction</a></h1>













IDEA (Interactive Display for Evolutionary Analyses) provides
a graphical interface for <a href="http://abacus.gene.ucl.ac.uk/software/paml.html" target="_top">PAML</a> (Phylogenetic Analysis by
Maximum Likelihood, Ziheng Yang, 1997), a suite of programs for
conducting molecular evolution analyses on nucleotide and amino-acid
data. &nbsp;IDEA allows you to run
either of the PAML programs <span style="font-style: italic;">codeml</span>
or <span style="font-style: italic;">baseml</span>
on
a single dataset or on multiple datasets simultaneously.
&nbsp;Among other things, <span style="font-style: italic;">codeml</span>
and <span style="font-style: italic;">baseml</span>
allow you to obtain maximum
likelihood estimates of numbers of substitutions per branch and per
site and to compare multiple models of molecular evolution given the
data and a phylogenetic tree for the sequences.<br>













<br>













<span style="font-weight: bold; font-style: italic; text-decoration: underline;">IDEA
allows you to:</span><br>













&nbsp;<br>













&bull; Run either of the&nbsp;<a href="http://abacus.gene.ucl.ac.uk/software/paml.html" target="_top">PAML</a> programs <span style="font-style: italic;">codeml</span> or <span style="font-style: italic;">baseml</span>. <br>













&bull; Optionally generate phylogenetic trees with <a href="http://atgc.lirmm.fr/phyml/">PhyML</a>, using maximum likelihood, or with
<a href="http://evolution.genetics.washington.edu/phylip.html" target="_top">PHYLIP</a>, using maximum parsimony
(small datasets) or neighbor-joining (large datasets). <br>













&bull;&nbsp;Perform multiple runs of <span style="font-style: italic;">codeml</span> with
different
starting <span style="font-style: italic;">&omega;</span>
(<span style="font-style: italic;">d</span>N/<span style="font-style: italic;">d</span>S)
values and merge their results for increased accuracy. <br>













&bull;&nbsp;Save parameter values for future use. <br>













&bull;&nbsp;Analyze many datasets in parallel. <br>













&bull;&nbsp;Monitor progress step by step.<br>













<br>













<span style="font-weight: bold; text-decoration: underline;"><i>If
you run a `site models' analysis (</i><span style="font-family: monospace;">seqtype = 1</span>
<i>, </i><span style="font-family: monospace;">model
= 0</span>
<i>), you can use IDEA's graphical output interface to:</i></span><br>













&nbsp;<br>













&bull;&nbsp;Easily identify genes and sites under positive
selection. <br>













&bull;&nbsp;Visualize selective pressure along a gene. <br>













<br>













IDEA is maintained by <a href="mailto:aegan@som.umaryland.edu">Amy
Egan</a>.<br>













<br>













<b>IDEA Research Team:</b>&nbsp;&nbsp;<a href="mailto:aegan@som.umaryland.edu">Amy
Egan</a> and <a href="mailto:jcsilva@som.umaryland.edu">Joana
C. Silva</a><br>













<br>













Inquiries may be posted to the <a href="http://sourceforge.net/forum/forum.php?forum_id=706804">IDEA
Support and Discussion Forum</a>.<br>













<br>













<hr>
<h1><a name="2">2&nbsp;</a>Running
IDEA</h1>













After IDEA has been <a href="download_idea.html#download">installed</a>,
it may be run by entering the IDEA installation directory (the
directory in which <span style="font-family: monospace; color: rgb(153, 0, 0);">idea-<span style="font-style: italic;">X</span>.<span style="font-style: italic;">X</span>.<span style="font-style: italic;">X</span>.tar.gz</span>
was saved or, equivalently, the directory in which the executable <span style="font-family: monospace; color: rgb(0, 0, 128);">idea</span>
is found) and typing <span style="font-family: monospace; color: rgb(0, 102, 0);">idea</span>
. &nbsp; For convenience, you may wish to add IDEA to your path;
doing so will allow you to run IDEA by simply typing <span style="font-family: monospace; color: rgb(0, 102, 0);">idea</span>
while in
any directory.<br>













<br>













The command for changing your path depends on
which shell you are using. &nbsp;Type <span style="font-family: monospace; color: rgb(0, 102, 0);">echo
$0</span> to find
out which shell you are using.<br>













<br>













In <span style="font-style: italic;">tcsh</span>:&nbsp;
<code><span style="color: rgb(0, 102, 0);">set PATH
= (${PATH}:</span><span style="color: rgb(255, 0, 0);">/your/IDEA/installation/directory</span>
<span style="color: rgb(0, 102, 0);">)</span></code><br>













In <span style="font-style: italic;">bash</span>:&nbsp;
<code><span style="color: rgb(0, 102, 0);">export
PATH=$PATH:</span><span style="color: rgb(255, 0, 0);">/your/IDEA/installation/directory</span></code><br>













<br>













To add IDEA to your path permanently, add the appropriate command above
to the <span style="font-family: monospace; color: rgb(0, 0, 128);">.cshrc</span>
or <span style="font-family: monospace; color: rgb(0, 0, 128);">.bashrc</span>
file in your home directory.<br>













<br>













<hr style="width: 100%; height: 2px;">
<h1><a name="3">3 Performing an Analysis</a></h1>













Before starting an analysis, you should be sure your input
data is in
the expected format.<br>













<h2><a name="3_1">3.1 Expected Input</a></h2>













<p>IDEA expects different input files depending on
whether you
have multiple datasets or a single dataset and whether you intend to
use existing phylogenetic trees. &nbsp;All these files should be in
the same directory (or directory hierarchy).<br>













</p>













<h3><a name="3_1_1">3.1.1 Alignment File(s)</a></h3>













For each dataset,&nbsp; IDEA expects a single nucleotide alignment
file in PAML format (here referred to as <span style="font-weight: bold;">PAMLseq</span>), which is a
gapped FASTA format with a header.<br>













<br>













The format of the header is<br>













<br>













&lt;space&gt; <span style="font-style: italic;">n</span>
&lt;space&gt; <span style="font-style: italic;">l</span><br>













<br>













, where <span style="font-style: italic;">n</span>
is
the number of sequences and <span style="font-style: italic;">l</span>
is the length of the nucleotide alignment.<br>













<br>













<span style="font-weight: bold;">Note:</span>
&nbsp;We recommend avoiding sequence names beginning with numbers.<br>













<br>













Below is an example PAMLseq alignment file.<br>













<br>













<code>&nbsp;5 537</code><br>













<code>&gt;species1</code><br>













<code>ATGGCATGTAAAGTTGATAAAGCTTTAGAGCATTCTACCCAAAATGAAGCACCCTCA---</code><br>













<code>---------AAAAATTATATGAACAATTTGTGTTATTACAAAAATAATGAATTAAAAAAA</code><br>













<code>ATAGACTCATCATATTTTCAAGATAAGTATTTAGGATTATTTTTTGGAGCTTCATGGTGT</code><br>













<code>AAATATTGTGTATCATTTATAAATAATTTGAATTTATTTAAAACCTACTTTCCCTTTTTT</code><br>













<code>GAAATAATATATATACCATTTGATCAAACATATACAGATTATATCAATTTTTTAAAAAAT</code><br>













<code>ACTAATTTTTATAGTTTACCTTTTGATAATTATTTATATATAGCTAATAAATTTAAAGTC</code><br>













<code>ACAAATTTGCCATCTTTTATTATTATAGCACCCAATAATAATATCCTTGTTAGGGATGGA</code><br>













<code>GTGCAATTAATTAAAACTGACAACTATATAAACAACTTCAAATCCTTGATAAAAAATTAT</code><br>













<code>ACAATACACCCCAAAACATTCAAATCAAATAATCGATTTTTCGATTTATTCTACAAT</code><br>













<code>&gt;species2</code><br>













<code>ATGGCTTGTAAAGTTGATAAAGCTCCAGAGCATCCTACCCAAAATGAAGTACCCTCA---</code><br>













<code>---------CAAAATTATATGAACAATTTATGTTATTACAAAAATAATGAATTAAAAAAA</code><br>













<code>ATAGACTCATCATATTTTCAAGATAAATATTTAGGATTATTTTTTGGAGCCTCATGGTGT</code><br>













<code>AAATATTGTGTATCATTTATAAATAATTTGAATTTATTTAAAACCTACTTCCCCTTCTTT</code><br>













<code>GAAATCATATATATACCATTTGATCAAACATATACAGATTATATTAATTTTTTAAAAAAT</code><br>













<code>ACAAATTTCTATAGCTTACCTTTTGATAATTATTTATATATAGCTAATAAATTTAAAGTC</code><br>













<code>AAAAATTTACCATCATTTATTATTATAGCACCCAATAATAATATCTTG---AGGGATGGT</code><br>













<code>GTGCAATTAATTAAAACAGACACCTATCTAAATAATTTCAAATCATTGATAAAAAATTAT</code><br>













<code>ACAATACACCCAAAAACATTCAAATCAAATAACCGATTTTTCGACTTATTCTACAAT</code><br>













<code>&gt;species3</code><br>













<code>ATGGCATGTAAAGTTGATAAAGTTTTAGAGCATCCTACCCAAAATGAAGAAACCTCA---</code><br>













<code>---------AAAAATTATATGAACAATTTGTTTTATTACAAAAATAATGAATTAAAAAAA</code><br>













<code>ATAGACTCATCATATTTTCAAGATAAATATTTAGGATTATTTTTTGGAGCCTCATGGTGT</code><br>













<code>AAATATTGTGTATCATTTATAAATAATTTAAATTTATTTAAAACTTATTTTCCTTTTTTT</code><br>













<code>GAAATTATATATATACCATTTGATCAAACATATACAGATTATATTAATTTTTTAAAAAAT</code><br>













<code>ACAAATTTTTATACTTTACCTTTTGATAATTATTTATATATAGCTAATAAATTTAAAGTC</code><br>













<code>AAAAATTTGCCATCTTTTATTATTATAGCACCAAATAATAATATACTTGTTAGGGATGGA</code><br>













<code>GTACAATTAATTAAAACTGACAATTATGTAAATAATTTCAAATCTTTGATAAAAAATTAT</code><br>













<code>ACAATACACCCCAAAACATTCAAATCAAATAATCGATTTTTCGACTTATTCTACAAT</code><br>













<code>&gt;species4</code><br>













<code>ATGGCGTGCCAAGTTGATAACCCCCCTAAAACATACCCAAACGATAAAACAGCTGAATAC</code><br>













<code>GAAAAGTACGCAAATTATATGAACTATCTATATTATTATCAAAATAATGAATTAAAAAAA</code><br>













<code>ATCGATTCCTCTTATTTTAAAGATAAATATTTAGGATTATTTTTTGGAGCTTCATGGTGT</code><br>













<code>AAATACTGTGTAACCTTTATAGATAGCTTAAATATATTTAAAAAGAACTTCCCCAATGTT</code><br>













<code>GAAATTATATATATACCATTTGATAGAACATATCAAGAGTACCAATCCTTTTTAAAAAAT</code><br>













<code>ACAAACTTTTATGCTTTACCTTTTGATAATTATTTATATATATGTAAAAAGTATCAAATA</code><br>













<code>AAAAATCTACCTTCCTTTATGTTAATTACACCTAATAATAATATACTAGTAAAGGATGCA</code><br>













<code>GCACAATTAATTAAAACAGATGAATATATAAATAATTTAAAATCATTAATAAAAAATTAT</code><br>













<code>ATCATACATCCTAAAACGTTTCAATTTAATAATCGCTTTTTTGATTTGTTTCGTAAT</code><br>













<code>&gt;species5</code><br>













<code>ATGAAATGCCAAGTGGATCGCCCCGTTACACCAAACGAAGAGCTAAATGGGGGCCAACAA</code><br>













<code>AATGTAGCCAAAAATTACATCCCCCATTTGTATCAATTCCAAAATAATGAAATGAAAAAA</code><br>













<code>ATCGATGCGTCTTACTTTGATAATAAATATCTGGGGCTATTTTTTGGAGCATCCTGGTGC</code><br>













<code>AGGTATTGCGTAACTTTCATCCAAAAAATAAATTTTTTTAAAAAGAATTTCCCCTTTATA</code><br>













<code>GAAATTATATACATCCCTTTTGACAAGACATATAATGATTATATAGCTTTCCTAAAAGGG</code><br>













<code>ACCGACTTTTACAGCCTTCCTTTTGATAACTATCTCTACGTTTGCAAAAAATTTAATGTT</code><br>













<code>CAAAATTTGCCATCCTTTATGATCATAGCCCCCAACAACAATGTGCTCGTCAAGGATGCC</code><br>













<code>GTGCAGCTCATCAAGACGGATGCCTACGTGGCGAACTTCAAGTCGTTGGTGAAAAATTAC</code><br>













<code>ACAATTCACCCGAACCAGTTTAAGTTTGGCAACCGATTTTTCGACTTATTTTGCGCA</code><br>













<h3><a name="3_1_2">3.1.2&nbsp;Tree
File(s)</a></h3>




Phylogenetic trees are required for an analysis.&nbsp; IDEA can
create these tree files for you using PhyML or PHYLIP (assuming the
desired program is installed on your system).&nbsp;&nbsp;PhyML estimates phylogenies using a maximum
likelihood approach. &nbsp;When using PHYLIP, phylogenetic trees will
be estimated using
maximum parsimony if six or fewer taxa are present. Otherwise, the
neighbor-joining algorithm will be used. If multiple trees are found,
the first one reported is used as input for PAML.<br>













<br>













If you prefer to use
trees that you have already created, your tree(s) should be in unrooted
format.&nbsp; If you are running IDEA on multiple datasets, you may
have a separate tree file for each dataset.
<h3><a name="3_1_3">3.1.3&nbsp;Dataset
List</a></h3>













When running IDEA on multiple datasets, an additional file is
required.
&nbsp;This file should list the name of each PAMLseq input file,
one filename per line. &nbsp;Below is an example dataset list:<br>













<br>













<code>gene1.PAMLseq</code><br style="font-family: monospace;">













<code>gene2.PAMLseq</code><br style="font-family: monospace;">













<code>gene3.PAMLseq</code><br>













<br>













If you intend to use existing phylogenetic trees, your dataset list
should contain a second column listing the name of the tree file for
each dataset. &nbsp;The first and second columns should be
separated by a tab. &nbsp;Below is an example dataset list that
includes tree filenames:<br>













<br>













<code>gene1.PAMLseq&nbsp;&nbsp;&nbsp;
gene1.tree</code><br style="font-family: monospace;">













<code>gene2.PAMLseq&nbsp;&nbsp;&nbsp;
gene2.tree</code><br style="font-family: monospace;">













<code>gene3.PAMLseq&nbsp;&nbsp;&nbsp;
gene3.tree</code><br>













<br>













Note that IDEA interprets the alignment and tree filenames you specify
as relative to your input directory. &nbsp;They may be organized in
subdirectories if you prefer. &nbsp;Below is an example that uses
subdirectories:<br>













<br>













<code>set1/alignments/gene1.PAMLseq&nbsp;&nbsp;&nbsp;
set1/trees/gene1.tree</code><br style="font-family: monospace;">













<code>set1/alignments/gene2.PAMLseq&nbsp;&nbsp;&nbsp;
set1/trees/gene2.tree</code><br style="font-family: monospace;">













<code>set2/alignments/gene3.PAMLseq&nbsp;&nbsp;&nbsp;
set2/trees/gene3.tree</code><br style="font-family: monospace;">













<br>













<h2><a name="3_2">3.2 Setting Parameters</a></h2>













Once you have prepared your input
files, you are ready to begin your analysis by running the command <code>idea</code> .
&nbsp;You should see the following window:<br>













<br>













<p align="center"><strong>Fig. 1:</strong>
IDEA Start Page<br>













<img style="border: 1px solid ; width: 339px; height: 256px;" alt="IDEA Start Page" src="introPage.jpg" align="middle"></p>













<br>













After you choose "Start New Analysis", you will have the option to set
multiple parameters on subsequent pages. &nbsp;Some of these
parameters are described below.<br>













<br>













<br>













<table style="width: 755px; height: 58px; text-align: left; margin-left: auto; margin-right: auto;" border="1" cellpadding="2" cellspacing="2">













  <tbody>













    <tr>













      <td><span style="font-weight: bold;">PAML
program</span></td>













      <td style="width: 80%;">Choose <span style="font-style: italic;">codeml</span> to perform
codon- or amino-acid-based analysis.<br>













Choose <span style="font-style: italic;">baseml</span>
to perform nucleotide-based analysis.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">IDEA mode</span></td>













      <td>Multi-dataset mode is the preferred method for
analyzing multiple datasets.<br>













It uses grid resources to execute many jobs simultaneously.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">Dataset
name list</span><br>













(multi-dataset mode)</td>













      <td>The name of the dataset list file described in section <a href="#3_1_3">3.1.3</a>.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">Input
directory</span><br>













(multi-dataset mode)</td>













      <td>This directory should contain all the files described
in section <a href="#3_1">3.1</a>,<br>













      <span style="font-style: italic;">except</span>
possibly the dataset list.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">Output
directory</span><br>













(multi-dataset mode)</td>













      <td>This is the directory where all output of an IDEA run
will go.<br>













It is recommended that this not be the same as the input directory.<br>













If you specify a directory that does not already exist, IDEA will
attempt to create it for you.<br>













It is recommended that no other files be stored in this directory.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">runmode</span></td>













      <td>If you have only two sequences for each dataset and are
running <span style="font-style: italic;">codeml</span>,
set&nbsp;<code>runmode</code>
to -2 to perform a pairwise analysis. &nbsp;This will save
computation time and allow you to
view a specialized
output display.<br>













      <br>













If your dataset list includes both pairwise and non-pairwise datasets,
setting <code>runmode</code>
to -2 will result in a pairwise analysis for the pairwise datasets and
a standard analysis for the non-pairwise datasets.<br>













      <br>













Setting <code>runmode</code>
to 0 will always result in a standard analysis for all datasets.<br>













      <br>













Extra&nbsp;<span style="font-style: italic;">&omega;</span>
values are ignored in pairwise mode, as are certain PAML parameters.</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">Extra
omega values</span><br>













(<span style="font-style: italic;">codeml</span>)</td>













      <td>This IDEA feature allows you to specify additional
starting values for&nbsp;<span style="font-style: italic;">&omega;</span>.<br>













Separate multiple values with spaces.<br>













PAML will be run once for each starting value,<br>













including the first starting value given as the parameter <code>omega</code> in the
left column.<br>













Afterwards, IDEA will choose the best run for each dataset and
evolution model.<br>













Extra&nbsp;<span style="font-style: italic;">&omega;</span>
values are ignored in pairwise mode (<code>runmode</code> -2).</td>













    </tr>













    <tr>













      <td><span style="font-weight: bold;">PAML
Parameters</span></td>













      <td>Mouse over the <img style="width: 13px; height: 13px;" alt="i" title="information icon" src="i_icon.gif"> icon next
to a parameter for a brief note on that parameter.<br>













For more detailed documentation, please refer to the PAML manual
(version <a href="http://abacus.gene.ucl.ac.uk/software/pamlDOC.pdf">4</a>
or&nbsp; <a href="http://bioportal.cgb.indiana.edu/docs/tools/paml/pamlDOC.pdf" target="_blank">3.15</a>).<br>













      </td>













    </tr>













  
  
  
  
  
  
  
  
  
  
  
  
  </tbody>
</table>













<br>













<br>













You may notice that your entries for some parameters are validated by
IDEA for your convenience. &nbsp;In particular, IDEA alerts you if
the
names of your sequence, tree and/or output files are too long (which
would cause PAML to fail) and checks for errors in
user-loaded configuration files. &nbsp;IDEA's input validation is
by no means
exhaustive, and you are encouraged to ensure that your parameter
settings are compatible with each other and with your data before
starting analysis.<br>













<br>













<h3><a name="3_2_1">3.2.1 Saving
Configuration</a></h3>













IDEA provides reasonable default values for most
parameters.
&nbsp;However, if you frequently perform similar analyses or wish
to repeat an analysis, it's a good idea to save your parameter choices
so you won't have to enter them again. &nbsp;To save the parameters
you've chosen, choose "Save Configuration..." from the <span style="font-weight: bold;">File</span> menu. &nbsp;Your
file will be
written
in PAML's .ctl
format. &nbsp;This is compatible with running PAML
outside of IDEA, but you will lose all non-PAML parameters, including
directories and extra&nbsp;<span style="font-style: italic;">&omega;</span>
values.<br>













<br>













It is recommended that you save your configuration prior to starting
your run.<br>













<h3><a name="3_2_2">3.2.2
Loading
Configuration</a></h3>













You may load parameters
from a
configuration file you saved
previously
or from any <code>codeml.ctl</code>
or <code>baseml.ctl</code>
file you already have. &nbsp;To do so, choose "Load
configuration..." in the <span style="font-weight: bold;">File</span>
menu. &nbsp;After loading your configuration, you may modify
parameters as you see fit.<br>













<h2><a name="3_3">3.3
Monitoring Progress</a></h2>













When you are satisfied
with your
parameter choices, press the <span style="font-weight: bold;">Start</span>
button.<br>













<span style="color: rgb(255, 0, 0);">You may not change
the parameters you have chosen after analysis has started.</span><br>













<br>













IDEA will use the <a href="http://sourceforge.net/projects/tigr-workflow/" target="_top">Workflow</a>
system to distribute your jobs over the grid. &nbsp;For each
dataset, IDEA may perform the following steps:<br>













<br>













&bull; Create a tree using PhyML or PHYLIP. &nbsp;This step will be
skipped if you have chosen to use your own trees.<br>













&bull; Run either <span style="font-style: italic;">codeml</span>
or <span style="font-style: italic;">baseml</span>.<br>













&bull; Merge separate runs of <span style="font-style: italic;">codeml</span>
if you gave multiple
starting&nbsp;<span style="font-style: italic;">&omega;</span>
values.<br>













&bull; Produce a tabular text summary of results. &nbsp;This
file will be located in the output directory you specified and will
have the suffix <code>.summary</code>&nbsp;.
&nbsp;If you are running <span style="font-style: italic;">codeml</span>
and chose at least one nested pair of <code>NSsites</code>
models, an additional file with the suffix <code>.lrt</code> will be
produced. &nbsp;Its contents will be the results of a likelihood
ratio test performed by running the PAML program <code>chi2</code>.<br>













<br>













To monitor the progress of your analysis, click the <span style="font-weight: bold;">Monitor</span> button.
&nbsp;This will bring up a <a href="http://tigr-workflow.sourceforge.net/user_guide.html#7_8"><span style="font-style: italic;">Workflow
monitor</span></a> in a separate
window. &nbsp;A process's output may be viewed by selecting the
process in the tree at left and clicking the&nbsp;<span style="font-weight: bold;">View Out</span> button (for
standard output) or the&nbsp;<span style="font-weight: bold;">View
Err</span>
button (for error output) at right. &nbsp;Because there is a time
lag
in relaying this information to the monitor, in-monitor output viewing
is most useful for multi-dataset IDEA runs.<br>













<br>













You may close the monitor window at any time. &nbsp;However, do
not close the main IDEA window until your jobs have finished. &nbsp;<span style="font-weight: bold;">Doing so will cause your analysis
to be aborted.<br>













</span><br>













<span style="font-weight: bold;">Note:</span>
&nbsp;When analyzing a very large number of datasets, the Workflow
monitor may become unresponsive. &nbsp;This has no impact on the
running jobs. &nbsp;In this case, you may close the monitor or
leave it open. &nbsp;<span style="font-weight: bold;">However,
do not close the main IDEA window until your jobs have finished, or
your analysis will be aborted.</span><br>













<h2><a name="3_4">3.4
Troubleshooting</a></h2>













At times, one or more
steps of the
IDEA pipeline may fail or
may not
produce the results you expect. &nbsp;To assist you in
troubleshooting a failed IDEA run, the console output of each step in
the pipeline is available in your output directory. &nbsp;The
output files are named as follows, where <span style="font-family: monospace; color: rgb(0, 102, 0);">&lt;dataset
name&gt;</span> is the name of an alignment file, or "ONLY"
in the case of a single-dataset analysis, and <span style="font-family: monospace; color: rgb(0, 102, 0);">yyyy-mm-dd</span>
is the date on which the analysis was started:<br>













<br>













<table style="text-align: left; width: 100%;" border="0" cellpadding="2" cellspacing="2">













  <tbody>













    <tr>













      <td style="width: 50%;">Standard output:</td>













      <td>Standard error:</td>













    </tr>













    <tr>













      <td><code>idea.create-tree.<span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.out.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.run-paml.<span style="color: rgb(0, 102, 0);">&lt;dataset
name&gt;</span>.out.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.merge-runs.<span style="color: rgb(0, 102, 0);">&lt;dataset
name&gt;</span>.out.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.parse-output.<span style="color: rgb(0, 102, 0);">&lt;dataset
name&gt;</span>.out.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span></code></td>













      <td><code>idea.create-tree.&lt;dataset
name&gt;.err.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.run-paml.&lt;dataset name&gt;.err.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.merge-runs.&lt;dataset name&gt;.err.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span><br>













idea.parse-output.&lt;dataset name&gt;.err.<span style="color: rgb(0, 102, 0);">yyyy-mm-dd</span></code></td>













    </tr>













  
  
  
  
  
  
  
  
  
  
  
  
  </tbody>
</table>













<br>













The log files <span style="font-family: monospace; color: rgb(0, 0, 128);">&lt;output
directory&gt;/idea.log</span> and/or <span style="font-family: monospace; color: rgb(0, 0, 128);">&lt;output
directory&gt;/workflow/idea.xml.log</span>, if present, may
also contain useful information.<br>













<br>













<span style="font-weight: bold;"></span>
<hr>
<h1><a name="4">4
Viewing Results</a></h1>













Once your analysis is
finished, you
will see a message
indicating
whether it succeeded. &nbsp;If so, click the <span style="font-weight: bold;">View Results</span> button
to get an interactive display of your results. &nbsp;This display
is intended for viewing the&nbsp;results of &lsquo;site
models&rsquo;<span style="font-style: italic;">codeml</span>
analyses (<code>seqtype = 1</code>;
<code>model = 0</code>),
although a small subset of its features are also applicable
to <span style="font-style: italic;">baseml</span>
analyses and certain other <span style="font-style: italic;">codeml</span>
analyses.<br>













<br>













To view results of a previously completed analysis or partial results
for a not-yet-completed analysis, choose "View
Results of
Previous Analysis" from the start page. &nbsp;Then choose "Load
Results" in the <span style="font-weight: bold;">File</span>
menu. &nbsp;Select the directory you gave as the <span style="font-weight: bold;">output directory</span> and
click <span style="font-weight: bold;">Load</span>.<br>













<br>













The interactive display has several components, as described below.<br>













<br>













<h2><a name="4_1">4.1
Data Table</a></h2>













The data table is the most
prominent display component and the
only
display component in the case of a pairwise analysis. &nbsp;It
displays
data for each dataset and evolutionary model in several columns:<br>













<br>













&bull; <span style="font-weight: bold;"></span><span style="font-weight: bold;">Dataset</span>:<br>













<div style="margin-left: 40px;">the
name of each
dataset<br>













This is shown as the name of the
corresponding alignment file as described in section <a href="#3_1_1">3.1.1</a>.<br>













</div>













<br>













&bull; <span style="font-weight: bold; font-style: italic;">n</span>
(non-pairwise mode only):<br>













<div style="margin-left: 40px;">the
number of
sequences/variants in a
dataset<br>













</div>













<br>













&bull; <span style="font-weight: bold;"></span><span style="font-weight: bold;">Model</span>:<br>













<div style="margin-left: 40px;">one
of several
models for
how&nbsp;<span style="font-style: italic;">&omega;</span>
is
allowed vary among sites<br>













This corresponds to PAML's <code>NSsites</code>
parameter.<br>













</div>













&nbsp;&nbsp;&nbsp;
<br>













&bull; <span style="font-weight: bold;">Likelihood
Score</span>:<br>













<div style="margin-left: 40px;">the
lnL score for
a
particular model, as computed by PAML<br>













If PAML was run with multiple starting <span style="font-style: italic;">&omega;</span> values,
the run that produced the best score is used.<br>













A less negative score indicates a likelier model. &nbsp;The
likeliest model for each dataset is shown in bold to make it easy to
identify. &nbsp;One can test the hypothesis that positive selection
is occurring by comparing a model that allows&nbsp;<span style="font-style: italic;">&omega;</span>
&gt; 1 for some sites with another that does not using a likelihood
ratio test.<br>













</div>













<br>













&bull; <span style="font-weight: bold;"></span><span style="font-weight: bold;">Tree Length</span>:<br>













<div style="margin-left: 40px;">the
estimated
number of
substitutions per site along all branches in the tree</div>













<br>













&bull; <span style="font-weight: bold;"><span style="font-style: italic;">d</span>N</span>
(pairwise mode only):<br>













<div style="margin-left: 40px;">the
number of
nonsynonymous substitutions per nonsynonymous site<br>













</div>













<br>













&bull; <span style="font-weight: bold;"></span><span style="font-weight: bold;"><span style="font-style: italic;">d</span>S</span>
(pairwise mode only):<br>













<div style="margin-left: 40px;">the
number of
synonymous
substitutions per synonymous site<br>













</div>













<br>













<span style="font-weight: bold;"></span>
&bull; <span style="font-weight: bold;"></span><span style="font-style: italic; font-weight: bold;"></span><span style="font-style: italic; font-weight: bold;">&omega;</span>:<br>













<div style="margin-left: 40px;"><span style="font-style: italic;">d</span>N/<span style="font-style: italic;">d</span>S, a measure of
selective pressure<br>













</div>













<br>













&bull; <span style="font-weight: bold; font-style: italic;">&kappa;</span>:<br>













<div style="margin-left: 40px;">ts/tv,
the
transition/transversion rate ratio<br>













</div>













<br>













The data may be sorted based on any column. &nbsp;Use the <img style="width: 15px; height: 15px;" alt="sort ascending globally" src="sortAscendingGlobal.png"> and <img style="width: 15px; height: 15px;" alt="sort descending globally" src="sortDescendingGlobal.png">
buttons to sort across datasets. &nbsp;Use the <img style="width: 15px; height: 15px;" alt="sort ascending" src="sortAscending.png"> and <img style="width: 15px; height: 15px;" alt="sort descending" src="sortDescending.png"> buttons to sort the models within
each dataset.<br>













<br>













<h3><a name="4_1_1">4.1.1
Histograms</a></h3>













For some columns, a
histogram (<img style="width: 15px; height: 15px;" alt="show histogram" src="showHistogram.png">) button
is also available.
&nbsp;Press this button to see an interactive histogram of the data
points in the corresponding column. &nbsp;In non-pairwise mode, you
will be prompted to select a single model of evolution on which to base
the histogram. &nbsp;Use the "Number of bins" slider to dynamically
adjust the number of bins between two and 50 (and thus adjust the bin
size). &nbsp;You can also switch models by selecting a different
model from the drop-down box at the top of the window. &nbsp;Press
the "Save" button to save the histogram as a JPEG image.<br>













<br>













<p align="center"><strong>Fig.
2:</strong>
Adjustable Histogram<br>













<img style="border: 1px solid ; width: 420px; height: 553px;" alt="Adjustable Histogram" src="histogram.png" align="middle"></p>













<h3><a name="4_1_2">4.1.2
Detailed Output</a></h3>













Although IDEA displays
many useful
values at a glance, you may
at times
wish to examine the more detailed output PAML provides. &nbsp;Click
on a model name for any dataset to view PAML's output for that dataset.
&nbsp;The output will be presented in a scrollable window that is
automatically scrolled to the start of the results for the model you
clicked on; you may view results for other models by scrolling up or
down. &nbsp;Note that you may move this window and continue to
perform other functions while viewing the detailed output.<br>













<h3><a name="4_1_3">4.1.3
Likelihood Ratio
Tests</a></h3>













The input parameter <code>NSsites</code>
is used to specify models that allow&nbsp;<span style="font-style: italic;">&omega;</span>
to vary across sites.
&nbsp;The
available choices include pairs of nested models which are identical
except that one allows some sites to have&nbsp;<span style="font-style: italic;">&omega;</span>
&gt; 1. &nbsp;Comparing the likelihood of the observed data
under each
model therefore amounts to a test of the hypothesis that some sites are
under positive selection. &nbsp;For each such pair you selected, a
likelihood ratio test should have been performed as part of the final
step of the IDEA pipeline. &nbsp;If the test is significant at the
5%
level, a <img style="width: 11px; height: 11px;" alt="check" src="check.png"> will be visible to the right of
the
likelihood score for the alternative (more complicated) model;
otherwise, an <img style="width: 11px; height: 11px;" alt="X" src="x.png"> will be visible in its place.
&nbsp;Mousing
over the&nbsp;<img style="width: 11px; height: 11px;" alt="check" src="check.png"> or <img style="width: 11px; height: 11px;" alt="X" src="x.png">
icon will bring up a display of the details of the likelihood
calculation similar to that below.<br>













<br>













<div style="text-align: center;"><strong>Fig.
3:</strong>
Likelihood Ratio Test Details<br>













</div>













<div style="text-align: center;"><img style="width: 344px; height: 163px;" alt="likelihood ratio test details" src="lrt.png"><br>













</div>













<h3><a name="4_1_4">4.1.4
Saving the Table</a></h3>













To save the table in
tabular text
format, choose "Save Table"
in the <span style="font-weight: bold;">File</span>
menu.<br>













<h2><a name="4_2">4.2
Selected Sites</a></h2>













<p>In non-pairwise
mode, the
bottom of the screen is
devoted to
the display of selected sites. &nbsp;Click on any row in the table
(except rows for model <span style="font-style: italic;">0-one-ratio</span>)
to bring up a display of selected sites for that dataset and evolution
model. &nbsp;You will see the amino acid sequence for that dataset.
&nbsp;Above each amino acid are one or more colored bars
representing the probability that&nbsp;<span style="font-style: italic;">&omega;</span> at that
site falls into each of the site classes allowed by the model.
&nbsp;(Although <span style="font-style: italic;">&omega;</span>
can really take on continuous values, PAML's models assume it has one
of several
discrete values.) &nbsp;A color key listing the&nbsp;<span style="font-style: italic;">&omega;</span> value
for each site class can be found at the top left of the display.
&nbsp;The height of each colored bar is proportional to the
probability that&nbsp;<span style="font-style: italic;">&omega;</span>
at that site is in the matching site class; probabilities sum to 1.
&nbsp;The most probable site class for each site is shown as a
colored square under the amino acid. &nbsp;Some models, such as <span style="font-style: italic;">2-PositiveSelection</span>,
<span style="font-style: italic;">3-discrete</span>
and <span style="font-style: italic;">8-beta&amp;w&gt;1</span>,
may have a site class for which&nbsp;<span style="font-style: italic;">&omega;</span>
&gt; 1 (indicating that positive selection is likely at certain
sites). &nbsp;In such cases, amino acids for which the most
likely&nbsp;<span style="font-style: italic;">&omega;</span>
value is &gt; 1 will be shown in white on black. &nbsp;The
numbers directly above the colored bars indicate position (in amino
acids) on the sequence.</p>













<p>The selected
sites display
will be based on PAML's
Bayes
empirical
Bayes analysis when it is available; it will be based on PAML's
Na&iuml;ve
empirical Bayes analysis otherwise. &nbsp;The analysis used will be
noted in the title of the selected sites display.</p>







<p>Occasionally, the selected sites cannot be visualized. &nbsp;This is usually
the result of errors during the execution of PAML, which are in turn
usually caused by errors in the input data. &nbsp;In the case of such a
failure, you are given the opportunity to browse the text output of the
display-creation command in order to find the error.</p>













<p><span style="font-weight: bold;">Note:</span>
&nbsp;<span style="font-style: italic;">&omega;</span>
values &ge;&nbsp;100 for site classes may be shown incorrectly
in the color key because&nbsp;<span style="font-style: italic;">&omega;</span>
values&nbsp;&ge; 1000 may be truncated in PAML's output (for
example, an&nbsp;<span style="font-style: italic;">&omega;</span>&nbsp;value
of 1100 for a site class could be truncated to 100). &nbsp;This is
a
known problem in PAML and does not affect the overall&nbsp;<span style="font-style: italic;">&omega;</span> values
listed in the data table.</p>













The portion of the code
that
creates this picture was
contributed by <a href="mailto:jcrabtree@som.umaryland.edu">Jonathan
Crabtree</a>.<br>













<h2><a name="4_3">4.3
Phylogenetic Trees</a></h2>













<p>In non-pairwise
mode, the
right side of the screen is
devoted
to the display of selected sites. &nbsp;Click on the <span style="font-weight: bold;">Tree</span> button in
any row to display a phylogenetic tree for the selected dataset based
on the selected model. &nbsp;As an alternative, you can select a
dataset and model from the drop-down boxes in the tree area and press
<span style="font-weight: bold;">Display</span>.
&nbsp;The tree displayed will be drawn to scale, using the
branch lengths estimated by PAML. &nbsp;To see a close-up view,
click "click to enlarge".</p>













The portion of the code
that
creates this picture was
contributed by <a href="mailto:jbadger@jcvi.org">Jonathan
Badger</a>.<br>













<h2><a name="4_4">4.4
Results Not Obtained
Using IDEA</a></h2>













IDEA's output display is
designed
to display output generated
by IDEA.
&nbsp;However, you may use it to display non-IDEA-generated PAML
output. &nbsp;To do so, place the following files in one directory:<br>













<br>













<table style="text-align: left; width: 100%;" border="1" cellpadding="2" cellspacing="2">













  <tbody>













    <tr>













      <td style="font-family: monospace; color: rgb(0, 0, 128);">list.txt</td>













      <td>This should be a list with the name of one sequence
file on each line.<br>













There may optionally be a second column containing the name of one tree
file for each<br>













sequence file.<br>













For the purposes of the output display, the sequence files listed in
list.txt need not actually exist.</td>













    </tr>













    <tr>













      <td style="font-family: monospace; color: rgb(0, 0, 128);">&lt;sequence
file name&gt;.lrt</td>













      <td>(Optional)<br>













One such file may be provided for each sequence file listed in <span style="font-family: monospace; color: rgb(0, 0, 128);">list.txt</span>.<br>













To view LRT results, generate these files by calling <span style="font-family: monospace; color: rgb(0, 96, 0);">idea-D-parse-output.pl</span>.</td>













    </tr>













    <tr>













      <td><span style="color: rgb(0, 0, 128); font-family: monospace;">sequence
file name&gt;.mlc</span><br>













-OR-<br>













      <span style="color: rgb(0, 0, 128); font-family: monospace;">&lt;sequence
file name&gt;.SINGLE.PAMLout.merged</span></td>













      <td>One such file is required for each sequence file listed
in <span style="font-family: monospace; color: rgb(0, 0, 128);">list.txt</span>.<br>













This should be a PAML output file.</td>













    </tr>













    <tr>













      <td style="font-family: monospace; color: rgb(0, 0, 128);">&lt;sequence
file name&gt;.mst</td>













      <td>To view selected sites, one such file is required for
each sequence file listed in <span style="font-family: monospace; color: rgb(0, 0, 128);">list.txt</span>.<br>













This should be an RST file as output by PAML.<br>













(The <code>.mst</code>
suffix reflects the fact that IDEA generates a merged RST file.)<span style="font-family: monospace; color: rgb(0, 0, 128);"></span><br>













      </td>













    </tr>













  
  
  
  
  
  
  
  
  
  
  
  
  </tbody>
</table>













<br>













<br>













Any of these files may be a symbolic link to a file in another location.<br>













<br>













Once you have placed all these files in one directory, simply select
that directory when loading results.<br>











<h2><a name="4_5">4.5 Examining Flat Files</a></h2>











Like any PAML analysis, an IDEA analysis results in traditional text
output files. &nbsp;At times, you may wish to examine these files.
&nbsp;Look for files with the following names:<br>











<br>











<table style="text-align: left; width: 100%;" border="0" cellpadding="2" cellspacing="2">











  <tbody>











    <tr>











      <td style="width: 346px;"><b>Multiple-dataset analyses:</b></td>











      <td style="width: 347px;"><b>Single-dataset analyses:</b></td>











    </tr>











    <tr>











      <td style="width: 346px;"><code><span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.PAMLtree<br>











      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.w<span style="color: rgb(0, 102, 0);">X_X</span>.codeml.ctl</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.w<span style="color: rgb(0, 102, 0);">X_X</span>.lnf</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.w<span style="color: rgb(0, 102, 0);">X_X</span>.rst</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.mst</code><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);">*<span style="color: rgb(0, 0, 0);"> (merged rst)</span></span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.w<span style="color: rgb(0, 102, 0);">X_X</span>.PAMLout</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.mlc</code><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);">* <span style="color: rgb(0, 0, 0);">(merged PAMLout)</span></span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.summary<br>











      <span style="color: rgb(0, 102, 0);">&lt;dataset name&gt;</span>.lrt<br>











      <span style="color: rgb(0, 102, 0);"></span></code></td>











      <td style="width: 347px;"><code>2NG.dN<br>











2NG.dS<br>











2NG.t<br>











MyDataset.PAMLtree<br>











ONLY.w<span style="color: rgb(0, 102, 0);">X_X</span>.codeml.ctl</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
ONLY.w<span style="color: rgb(0, 102, 0);">X_X</span>.lnf</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
ONLY.w<span style="color: rgb(0, 102, 0);">X_X</span>.rst</code><span style="color: rgb(255, 0, 0);">*</span><br>










      <code>
ONLY.mst</code><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);">*<span style="color: rgb(0, 0, 0);"> (merged rst)</span></span><br>










      <code>
rates<br>











rst1<br>











rub<br>











      <span style="color: rgb(0, 102, 0);"></span><span style="color: rgb(0, 102, 0);">&lt;output file name&gt;</span>.w<span style="color: rgb(0, 102, 0);">X_X</span></code><span style="color: rgb(255, 0, 0);">*</span><code><span style="color: rgb(0, 102, 0);"></span><br>











      <span style="color: rgb(0, 102, 0);">&lt;output file name&gt;</span>.merged</code><span style="color: rgb(255, 0, 0);"></span><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(0, 0, 0);"> (merged PAMLout)</span><br>










      <code>
      <span style="color: rgb(0, 102, 0);">&lt;output file name&gt;</span>.summary<br>











      </code><code><span style="color: rgb(0, 102, 0);">&lt;output file name&gt;</span>.lrt</code><br>











      </td>











    </tr>











  
  
  
  
  
  
  
  
  
  
  
  </tbody>
</table>











<br>











<span style="color: rgb(255, 0, 0);"><span style="color: rgb(0, 0, 0);"></span>*</span>
These files are parameterized by&nbsp;<span style="font-style: italic;">&omega;</span>; there is one copy of each of these files for each starting&nbsp;<span style="font-style: italic;">&omega;</span> value. &nbsp;For example, the file <span style="font-family: monospace;">ONLY.w1_5.rst</span> would correspond to the starting&nbsp;<span style="font-style: italic;">&omega;</span> value 1.5.<br>










<span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);">*</span><span style="color: rgb(255, 0, 0);"></span>
These files contain the merged results of runs with different starting&nbsp;<span style="font-style: italic;">&omega;</span> values.<br>










<br>













<hr style="width: 100%; height: 2px;">
<h1>5 <a name="5"></a>New
in IDEA 2.5</h1>

New features in IDEA 2.5 include:<br>













<br>













&bull; <span style="font-weight: bold;">PhyML</span>:<br>













<div style="margin-left: 40px;">PhyML has now replaced PHYLIP as the
standard software for creating trees. &nbsp;Note that PhyML is not
supported on Solaris. &nbsp;PHYLIP is still available as an alternative.<br>













</div>













<span style="font-weight: bold;"></span><br>













&bull;<span style="font-weight: bold;"></span><span style="font-weight: bold;"></span><span style="font-weight: bold;"> Support for PAML 4.3 and 4.4</span><span style="font-weight: bold;"></span><span style="font-weight: bold;"></span><span style="font-weight: bold;"></span><br>













<br>













<hr style="width: 100%; height: 2px;">
<h1>6&nbsp;<a name="6"></a>Further
Information</h1>













Please visit the <a href="http://sourceforge.net/forum/forum.php?forum_id=706804">IDEA
Support and Discussion Forum</a>
to request support, provide feedback and discuss IDEA with other users.
&nbsp;Using this public forum will help the IDEA research team
provide
you with a timely response.<br>













<h2><a name="4_1">6.1 How to Cite IDEA</a></h2>













<code>Egan A., A. Mahurkar, J.
Crabtree, J.H. Badger, J.M. Carlton and J.C. Silva.
</code><br style="font-family: monospace;">













<code>IDEA:&nbsp; Interactive Display for Evolutionary Analyses.<br>








BMC Bioinformatics 2008, 9:524</code><code>.</code><br>








<br>








Please remember that if you use IDEA, you should also cite <a href="http://abacus.gene.ucl.ac.uk/software/paml.html">PAML</a>, <a href="http://www.atgc-montpellier.fr/phyml">PhyML</a> and <a href="http://evolution.genetics.washington.edu/phylip/faq.html#citation">PHYLIP</a>.<paml reference=""><phylip reference=""><br>













</phylip></paml>
</body>
</html>
